components/common-go/baseserver/server.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package baseserver
+
+import (
+	"context"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+func New(name string, opts ...Option) (*Server, error) {
+	cfg, err := evaluateOptions(defaultConfig(), opts...)
+	if err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	server := &Server{
+		Name: name,
+		cfg:  cfg,
+	}
+
+	if initErr := server.initializeHTTP(); initErr != nil {
+		return nil, fmt.Errorf("failed to initialize http server: %w", initErr)
+	}
+	if initErr := server.initializeGRPC(); initErr != nil {
+		return nil, fmt.Errorf("failed to initialize grpc server: %w", initErr)
+	}
+
+	return server, nil
+}
+
+type Server struct {
+	// Name is the name of this server, used for logging context
+	Name string
+
+	cfg *config
+
+	// http is an http Server
+	http         *http.Server
+	httpMux      *http.ServeMux
+	httpListener net.Listener
+
+	// grpc is a grpc Server
+	grpc         *grpc.Server
+	grpcListener net.Listener
+
+	// listening indicates the server is serving. When closed, the server is in the process of graceful termination.
+	listening chan struct{}
+}
+
+func (s *Server) ListenAndServe() error {
+	var err error
+	s.grpcListener, err = net.Listen("tcp", fmt.Sprintf(":%d", s.cfg.grpcPort))
+	if err != nil {
+		return fmt.Errorf("failed to acquire port %d", s.cfg.grpcPort)
+	}
+
+	s.httpListener, err = net.Listen("tcp", fmt.Sprintf(":%d", s.cfg.httpPort))
+	if err != nil {
+		return fmt.Errorf("failed to acquire port %d", s.cfg.grpcPort)
+	}
+
+	errors := make(chan error)
+	defer close(errors)
+	s.listening = make(chan struct{})
+
+	go func() {
+		s.Logger().
+			WithField("protocol", "grpc").
+			Infof("Serving gRPC on %s", s.grpcListener.Addr().String())
+		if serveErr := s.grpc.Serve(s.grpcListener); serveErr != nil {
+			if s.isClosing() {
+				return
+			}
+
+			errors <- serveErr
+		}
+	}()
```

Comment:
Why does this server (and the http one below) need to be started in a goroutine? Could we not simply error out if we fail to start either of them without having them send their errors on a channel only to collect them on line 105?

Reply:
Needs to be a go-routine because grpc.Serve and `http.ListenAndServe` are blocking. This follows the standard golang practice where if you want things to be in the background, the caller can do that themselves.

We need to put them in go-routines because we need to start both of them, and then synchronously await termination of either (or both) of the servers before exiting to maintain the semantics of listen and serve.